### PR TITLE
lightning: remove support for tikv-importer

### PR DESCRIPTION
**Pull request by gozssky**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/34310 https://github.com/pingcap/tidb/issues/31664

Problem Summary:

tikv-importer is out of maintenance since v5.0.6. tidb-lightning has also removed tikv-importer related documents since v6.0.

### What is changed and how it works?

Remove tikv-importer related logic and tests.

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [x] Breaking backward compatibility

Documentation

- [x] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
lightning: remove support for tikv-importer
```

### CODE REVIEW

**Review by ngaut — APPROVED** [review]

**Comment by gozssky** [comment]
Integration tests pass.

https://ci.pingcap.net/blue/organizations/jenkins/tidb_ghpr_integration_br_test/detail/tidb_ghpr_integration_br_test/8982/pipeline

File: br/pkg/lightning/config/global.go [hunk 1]
```diff
@@ -153,8 +152,7 @@ func LoadGlobalConfig(args []string, extraFlags func(*flag.FlagSet)) (*GlobalCon
 	tidbStatusPort := fs.Int("tidb-status", 0, "TiDB server status port (default 10080)")
 	pdAddr := fs.String("pd-urls", "", "PD endpoint address")
 	dataSrcPath := fs.String("d", "", "Directory of the dump to import")
-	importerAddr := fs.String("importer", "", "address (host:port) to connect to tikv-importer")
-	backend := flagext.ChoiceVar(fs, "backend", "", `delivery backend: local, importer, tidb`, "", "local", "importer", "tidb")
+	backend := flagext.ChoiceVar(fs, "backend", "", `delivery backend: local, importer, tidb`, "", "local", "tidb")
```
Comment (D3Hunter) [inline br/pkg/lightning/config/global.go:155]: remove "importer"?

File: br/pkg/lightning/restore/restore.go [hunk 2]
```diff
@@ -955,8 +944,6 @@ func (rc *Controller) saveStatusCheckpoint(ctx context.Context, tableName string
 
 // listenCheckpointUpdates will combine several checkpoints together to reduce database load.
 func (rc *Controller) listenCheckpointUpdates() {
-	defer rc.checkpointsWg.Done()
```
Comment (D3Hunter) [inline br/pkg/lightning/restore/restore.go:958]: why move this defer?

File: br/pkg/lightning/restore/restore.go [hunk 3]
```diff
@@ -2245,15 +2240,21 @@ func (cr *chunkRestore) deliverLoop(
 		// Update the table, and save a checkpoint.
 		// (the write to the importer is effective immediately, thus update these here)
 		// No need to apply a lock since this is the only thread updating `cr.chunk.**`.
-		// In local mode, we should write these checkpoint after engine flushed.
+		// In local mode, we should write these checkpoints after engine flushed.
 		cr.chunk.Checksum.Add(&dataChecksum)
 		cr.chunk.Checksum.Add(&indexChecksum)
 		cr.chunk.Chunk.Offset = currOffset
 		cr.chunk.Chunk.PrevRowIDMax = rowID
+		// currOffset may be less than endOffset even if we have read all rows of this chunk.
```
Comment (D3Hunter) [inline br/pkg/lightning/restore/restore.go:2248]: when will this happen? seems a bug

File: br/pkg/lightning/config/global.go (hunk 1 above)
Comment (gozssky) [reply br/pkg/lightning/config/global.go:155]: yes

File: br/pkg/lightning/restore/restore.go (hunk 2 above)
Comment (gozssky) [reply br/pkg/lightning/restore/restore.go:958]: failpoint will call `checkpointWg.Done` manually. This is for avoid panic like `panic: sync: negative WaitGroup counter`.

File: br/pkg/lightning/restore/restore.go (hunk 3 above)
Comment (gozssky) [reply br/pkg/lightning/restore/restore.go:2248]:
```
insert into t(a) values (1, 2, 3);
```
`currOffset` will be 
```
insert into t(a) values (1,2,3);
                              ^
                              |
```

Comment (gozssky) [reply br/pkg/lightning/restore/restore.go:2248]: In other words, reading all rows isn't equivalent to reading all bytes of a chunk.

File: br/pkg/lightning/config/global.go (hunk 1 above)
Comment (gozssky) [inline br/pkg/lightning/config/global.go:155]:
```suggestion
	backend := flagext.ChoiceVar(fs, "backend", "", `delivery backend: local, tidb`, "", "local", "tidb")
```

File: br/pkg/lightning/restore/restore.go (hunk 3 above)
Comment (D3Hunter) [reply br/pkg/lightning/restore/restore.go:2248]: encode loop may return error and then this channel is closed, in this case we should not do `cr.chunk.Chunk.Offset = cr.chunk.Chunk.EndOffset`, `channelClosed` is not enough to cover this case.

File: br/pkg/lightning/restore/restore.go (hunk 2 above)
Comment (D3Hunter) [reply br/pkg/lightning/restore/restore.go:958]: maybe add a comment, it violate normal convention to wait goroutine.

File: br/pkg/lightning/restore/restore.go (hunk 3 above)
Comment (gozssky) [reply br/pkg/lightning/restore/restore.go:2248]: Yes, you are right. 

`encodeLoop` will send a nil slice to `kvsCh` after reading all rows. `deliverLoop` can distinguish between channel closed and rows are all read.

File: br/pkg/lightning/restore/restore.go (hunk 2 above)
Comment (gozssky) [reply br/pkg/lightning/restore/restore.go:958]: Added.

File: br/pkg/lightning/restore/restore.go (hunk 3 above)
Comment (gozssky) [reply br/pkg/lightning/restore/restore.go]: Fixed. PTAL again, thanks.

**Review by D3Hunter — APPROVED** [review]
lgtm

**Pull request merged**